qiskit/dagcircuit/dagdependency_v2.py

```diff
@@ -0,0 +1,672 @@
+            bit (Bit): The bit to locate.
+
+        Returns:
+            namedtuple(int, List[Tuple(Register, int)]): A 2-tuple. The first element (``index``)
+                contains the index at which the ``Bit`` can be found (in either
+                :obj:`~DAGDependencyV2.qubits`, :obj:`~DAGDependencyV2.clbits`, depending on its
+                type). The second element (``registers``) is a list of ``(register, index)``
+                pairs with an entry for each :obj:`~Register` in the circuit which contains the
+                :obj:`~Bit` (and the index in the :obj:`~Register` at which it can be found).
+
+          Raises:
+            DAGDependencyError: If the supplied :obj:`~Bit` was of an unknown type.
+            DAGDependencyError: If the supplied :obj:`~Bit` could not be found on the circuit.
+        """
+        try:
+            if isinstance(bit, Qubit):
+                return self._qubit_indices[bit]
+            elif isinstance(bit, Clbit):
+                return self._clbit_indices[bit]
+            else:
+                raise DAGDependencyError(f"Could not locate bit of unknown type: {type(bit)}")
+        except KeyError as err:
+            raise DAGDependencyError(
+                f"Could not locate provided bit: {bit}. Has it been added to the DAGDependency?"
+            ) from err
+
+    def apply_operation_back(self, operation, qargs=(), cargs=()):
+        """Add a DAGOpNode to the graph and update the edges.
+
+        Args:
+            operation (qiskit.circuit.Operation): operation as a quantum gate
+            qargs (list[~qiskit.circuit.Qubit]): list of qubits on which the operation acts
+            cargs (list[Clbit]): list of classical wires to attach to
+        """
+        new_node = DAGOpNode(
+            op=operation,
+            qargs=qargs,
+            cargs=cargs,
+            dag=self,
+        )
+        new_node._node_id = self._multi_graph.add_node(new_node)
+        self._update_edges(new_node)
+        self._increment_op(new_node.op)
+
+    def _update_edges(self, new_node):
+        """
+        Updates DagDependencyV2 by adding edges to the newly added node (new_node)
+        from the previously added nodes.
+        For each previously added node (prev_node), an edge from prev_node to new_node
+        is added if new_node is "reachable" from prev_node (this means that the two
+        nodes can be made adjacent by commuting them with other nodes), but the two nodes
+        themselves do not commute.
+        """
+
+        reachable = {}
+
+        # Analyze nodes in the reverse topological order.An improvement to the original
+        # algorithm is to consider only direct predecessors and to avoid constructing the
+        # lists of forward and backward reachable predecessors for every node when not required.
+        def _key(x):
+            return x.sort_key
+
+        toposort_list = rx.lexicographical_topological_sort(self._multi_graph, key=_key)
+
```

Comment:
I thought we already had a function for this.

Reply:
The method `rx.lexicographical_topological_sort` returns a list, which is changed to an iterator in `topological_nodes`. In order to use `reversed()` to create a reverse iterator, it needs to act on a list.

Reply:
Hmm, we can still write something like ``toposort_list = list(self.topological_nodes())``, or do you think that it's going be not as efficient as the approach above?

Reply:
Done in https://github.com/Qiskit/qiskit/pull/11764/commits/e1e5eb7201eb253e4aba1c139a5cdcec2533ce62.